ipynb/energy/EnergyModel_SystemEnergy.ipynb

```diff
@@ -452,404 +457,1647 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## Active States Profiling"
+    "## Power performance analysis"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": 17,
+   "execution_count": 15,
    "metadata": {
-    "collapsed": false,
-    "scrolled": true
+    "collapsed": true
    },
    "outputs": [],
    "source": [
-    "def compute_power_perf(clusters, loop_cnt, benchmark, bkp_file='pstates.csv'):\n",
-    "    \"\"\"\n",
-    "    Perform P-States profiling on each input cluster.\n",
+    "class power_perf_recorder(object):\n",
     "    \n",
-    "    This method requires a `benchmark` object with the following\n",
-    "    characteristics:\n",
+    "    def __init__(self, clusters, benchmark, channel):\n",
+    "        self.clusters = clusters\n",
+    "        self.benchmark = benchmark\n",
+    "        self.channel = channel\n",
     "        \n",
-    "        - duration, attribute that tells the workload duration in seconds\n",
-    "        - run(cgroup, threads), run the benchmark into the specified 'cgroup',\n",
-    "                                spawning the specified number of 'threads',\n",
-    "                                and return a performance score of their execution.\n",
-    "                                \n",
-    "    Data will be saved into a CSV file at each iteration such that, if something\n",
-    "    goes wrong, the user can restart the experiment considering only idle_states\n",
-    "    that had not yet been profiled.\n",
-    "    \n",
-    "    :param clusters: list of clusters to profile\n",
-    "    :type clusters: list(namedtuple(ClusterDescription))\n",
-    "    \n",
-    "    :param loop_cnt: number of iterations for each experiment\n",
-    "    :type loop_cnt: int\n",
-    "    \n",
-    "    :param benchmark: benchmark object\n",
-    "    :type benchmark: int\n",
-    "    \n",
-    "    :param bkp_file: CSV file name\n",
-    "    :type bkp_file: str\n",
-    "    \"\"\"\n",
-    "\n",
-    "    # Make sure all CPUs are online\n",
-    "    target.hotplug.online_all()\n",
+    "    @property\n",
+    "    def online_cpus(self):\n",
+    "        return target.list_online_cpus()\n",
+    "        \n",
+    "    def _read_energy(self):\n",
+    "        energy = 0.0\n",
+    "        perf = { cpu : 0.0 for cpu in self.online_cpus }\n",
+    "\n",
+    "        for i in xrange(self.loop_cnt):\n",
+    "            te.emeter.reset()\n",
+    "\n",
+    "            for cpu, value in self.benchmark.run(self.online_cpus).iteritems():\n",
+    "                perf[cpu] += value\n",
+    "\n",
+    "            nrg = te.emeter.report(te.res_dir).channels\n",
+    "            energy += float(nrg[self.channel])\n",
+    "            \n",
+    "            # Per-iteration delay\n",
+    "            if self.delay > 0:\n",
+    "                sleep(self.delay)\n",
```

Comment:
It is, but I think it's better to ask the question: "do I need to sleep ?" rather than do "give the sleep function whatever argument I have and it'll take care of it for me". That's me being pedantic though.